### PR TITLE
[Notifier] add support for smsapi-notifier

### PR DESCRIPTION
**Pull request by szepczynski**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| License       | MIT
| Doc PR        | symfony/symfony-docs#13716

Hi,

I've created integration to notifier to support polish sms operator - smsapi.pl

Here is smsapi-notifier integration: https://github.com/szepczynski/smsapi-notifier. Can you grab this code and make as symfony/smsapi-notifier?

This PR includes changes in notifier and framework-bundle to support smsapi transport as well as other included in notifier component.

Could someone integrate this into notifier component?

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
Since this is a new feature, can you please rebase to target master?

**Comment by szepczynski** [comment]
ok, done

I create also a recipe PR here: https://github.com/symfony/recipes/pull/776

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/notifier_transports.xml [hunk 1]
```diff
@@ -50,6 +50,10 @@
             <tag name="texter.transport_factory" />
         </service>
 
+        <service id="notifier.transport_factory.smsapi" class="Symfony\Component\Notifier\Bridge\Smsapi\SmsapiTransportFactory" parent="notifier.transport_factory.abstract">
+            <tag name="texter.transport_factory" />
+        </service>
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/notifier_transports.xml:55]: Can you move it to the new PHP config file?

**Review by fabpot — CHANGES_REQUESTED** [review]
The code should be part of the PR. Can you add it here?

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/notifier_transports.xml (hunk 1 above)
Comment (szepczynski) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/notifier_transports.xml]: Yes, I've done

**Comment by fabpot** [comment]
You should rebase to get rid of the merge commit.

File: src/Symfony/Component/Notifier/Bridge/Smsapi/.gitattributes [hunk 2]
```diff
@@ -0,0 +1,3 @@
+/Tests export-ignore
+/phpunit.xml.dist export-ignore
+/.gitignore export-ignore
```
Comment (fabpot) [inline src/Symfony/Component/Notifier/Bridge/Smsapi/.gitattributes:3]: missing new line.

File: src/Symfony/Component/Notifier/Bridge/Smsapi/README.md [hunk 3]
```diff
@@ -0,0 +1,4 @@
+SMSAPI Notifier
+===============
+
+Provides Smsapi integration for Symfony Notifier.
```
Comment (fabpot) [inline src/Symfony/Component/Notifier/Bridge/Smsapi/README.md:4]: Can you add the DSN pattern as well? And the links at the end like done for the other norifiers?

File: src/Symfony/Component/Notifier/Bridge/Smsapi/SmsapiTransport.php [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    {
+        return sprintf('smsapi://%s?from=%s', $this->getEndpoint(), $this->from);
+    }
+
+    public function supports(MessageInterface $message): bool
+    {
+        return $message instanceof SmsMessage;
+    }
+
+    protected function doSend(MessageInterface $message): void
```
Comment (fabpot) [inline src/Symfony/Component/Notifier/Bridge/Smsapi/SmsapiTransport.php:41]: As of 5.2, you must return a `SentMessage` instance.

File: src/Symfony/Component/Notifier/Bridge/Smsapi/SmsapiTransportFactory.php [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+use Symfony\Component\Notifier\Exception\IncompleteDsnException;
+use Symfony\Component\Notifier\Exception\UnsupportedSchemeException;
+use Symfony\Component\Notifier\Transport\AbstractTransportFactory;
+use Symfony\Component\Notifier\Transport\Dsn;
+use Symfony\Component\Notifier\Transport\TransportInterface;
+
+/**
+ * @author Marcin Szepczynski <szepczynski@gmail.com>
+ */
```
Comment (fabpot) [inline src/Symfony/Component/Notifier/Bridge/Smsapi/SmsapiTransportFactory.php:13]: You should add `@experimental in 5.2` in all new PHP files.

File: src/Symfony/Component/Notifier/Bridge/Smsapi/SmsapiTransportFactory.php [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    public function create(Dsn $dsn): TransportInterface
+    {
+        $scheme = $dsn->getScheme();
+        $authToken = $dsn->getUser();
+        $host = 'default' === $dsn->getHost() ? null : $dsn->getHost();
+        $from = $dsn->getOption('from');
+        $port = $dsn->getPort();
+
+        if (!$authToken) {
+            throw new IncompleteDsnException('Missing path (maybe you haven\'t update the DSN when upgrading from 5.0).', $dsn->getOriginalDsn());
```
Comment (fabpot) [inline src/Symfony/Component/Notifier/Bridge/Smsapi/SmsapiTransportFactory.php:28]: I think this is a copy/paste that should be removed as there is no upgrade here as the provider is new.

**Review by fabpot — CHANGES_REQUESTED** [review]

**Review by fabpot — APPROVED** [review]
LGTM. Can you rebase to get rid of the merge commit?

**Comment by fabpot** [comment]
Can you also fix issues reported by fabbot (please, ignore the error message suggestion as this is a false positive)?

**Comment by fabpot** [comment]
Thank you @szepczynski.

**Pull request merged**